### PR TITLE
S3 list with key object

### PR DESCRIPTION
**Pull request by dlstadther**
## Description
Updated `S3Client.copy()` such that when copying a directory of files, the number and total size of files will be returned. To accomplish this, the `list()` and `listdir()` now have an optional parameter `return_key` which when set `True` will return the `boto.s3.key.Key` objects rather than the string filenames.

Module should remain backwards compatible.

## Motivation and Context
I need to perform a few dynamic calculations based on the number and size of copied files. To accomplish this, `S3Client.copy()` needs to return the number and size of files copied.

## Have you tested this? If so, how?
Tests are included.

### CODE REVIEW

**Comment by dlstadther** [comment]
@Tarrasch Does this satisfy your suggestion within #1741? If so, I'll merge Monday (EST).

File: luigi/s3.py [hunk 1]
```diff
@@ -449,7 +456,7 @@ def move(self, source_path, destination_path, **kwargs):
         self.copy(source_path, destination_path, **kwargs)
         self.remove(source_path)
 
-    def listdir(self, path, start_time=None, end_time=None):
+    def listdir(self, path, start_time=None, end_time=None, with_key=False):
         """
         Get an iterable with S3 folder contents.
         Iterable contains paths relative to queried path.
```
Comment (Tarrasch) [inline luigi/s3.py]: Maybe also add docs like `:param with_key: if True this function will instead return blah blah`?

**Comment by Tarrasch** [comment]
LGTM

**Comment by dlstadther** [comment]
@jpiper Since you've made the most recent s3 changes, could you take a look at this PR? Thanks

File: luigi/s3.py [hunk 2]
```diff
@@ -449,13 +458,14 @@ def move(self, source_path, destination_path, **kwargs):
...
 
         :param start_time: Optional argument to copy files with modified dates after start_time
         :param end_time: Optional argument to copy files with modified dates before end_time
+        :param with_key: Optional argument, when set to True will return a tuple (filename, boto_key_object)
```
Comment (jpiper) [inline luigi/s3.py]: Sorry to nitpick, but can you change this to something along the lines of

> Optional argument, when set to True causes function to yield (filename, boto.s3.key.Key) tuples

So we're a bit more explicit in the documentation - thanks!

File: luigi/s3.py [hunk 3]
```diff
@@ -472,12 +482,20 @@ def listdir(self, path, start_time=None, end_time=None):
...
         key_path_len = len(self._add_path_delimiter(path))
-        for item in self.listdir(path, start_time=start_time, end_time=end_time):
-            yield item[key_path_len:]
+        for path in self.listdir(path, start_time=start_time, end_time=end_time):
+            yield path[key_path_len:]
+
+    def list_with_key(self, path, start_time=None, end_time=None):
+        key_path_len = len(self._add_path_delimiter(path))
+        for path, item in self.listdir(path, start_time=start_time, end_time=end_time, with_key=True):
+            yield path[key_path_len:], item
```
Comment (jpiper) [inline luigi/s3.py]: My spidey senses tell me this could be condensed to a single function with an extra argument, but I'm not too concerned by it

**Comment by jpiper** [comment]
Apart from two minor nitpicks, this LGTM!

File: luigi/s3.py (hunk 2 above)
Comment (dlstadther) [inline luigi/s3.py]: You're right. I'll update this docstring

File: luigi/s3.py (hunk 3 above)
Comment (dlstadther) [inline luigi/s3.py]: Yeah... I couldn't think of a more eloquent way to combine them than the following:

```python
def list(self, path, start_time=None, end_time=None, with_key=False):
    key_path_len = len(self._add_path_delimiter(path))
    if with_key:
        for path, item in self.listdir(path, start_time=start_time, end_time=end_time, with_key=with_key):
            yield path[key_path_len:], item
    else:
        for path in self.listdir(path, start_time=start_time, end_time=end_time):
            yield path[key_path_len:]
```

The issue is that `listdir` return a generator of values when `with_key=False` and a generator of tuples when `with_key=True`. So I'm not sure there's a best way to handle these two generator formats with a single loop.

Perhaps, there's a better way to handle `listdir` when `with_key=True`. Possibly only return `boto.s3.key.Key` (such that listdir only yields a single value)? Thoughts?

**Comment by dlstadther** [comment]
@jpiper Thanks for reviewing so quickly!

**Comment by dlstadther** [comment]
@jpiper Could i get you to take one more look at this? I've changed it such that `list()` and `listdir()` will only ever return a single value from their generators - either a `str` or a `boto.s3.key.Key` object. I then updated `copy()` accordingly.

**Comment by dlstadther** [comment]
@jpiper @f355 @mikekap You all have had your hands in the `copy`, `list`, and `listdir` methods for which i have modified. Thoughts?

File: luigi/s3.py [hunk 4]
```diff
@@ -338,16 +340,23 @@ def copy(self, source_path, destination_path, threads=100, start_time=None, end_
...
+            total_size_bytes = 0
             src_prefix = self._add_path_delimiter(src_key)
             dst_prefix = self._add_path_delimiter(dst_key)
-            for key in self.list(source_path, start_time=start_time, end_time=end_time):
+            for item in self.list(source_path, start_time=start_time, end_time=end_time, return_key=True):
+                (bucket, key) = self._path_to_bucket_and_key(source_path)
+                key_path = self._add_path_delimiter(key)
+                key_path_len = len(key_path)
+                path = item.key[key_path_len:]
+                print(path)
```
Comment (f355) [inline luigi/s3.py]: Debugging leftover?

File: luigi/s3.py [hunk 5]
```diff
@@ -338,16 +340,23 @@ def copy(self, source_path, destination_path, threads=100, start_time=None, end_
...
             management_pool = ThreadPool(processes=threads)
 
+            total_size_bytes = 0
             src_prefix = self._add_path_delimiter(src_key)
             dst_prefix = self._add_path_delimiter(dst_key)
-            for key in self.list(source_path, start_time=start_time, end_time=end_time):
+            for item in self.list(source_path, start_time=start_time, end_time=end_time, return_key=True):
+                (bucket, key) = self._path_to_bucket_and_key(source_path)
+                key_path = self._add_path_delimiter(key)
+                key_path_len = len(key_path)
```
Comment (f355) [inline luigi/s3.py]: Three lines above can be moved outside of the `for` loop I think.

**Comment by f355** [comment]
Apart from that looks good!

File: luigi/s3.py (hunk 5 above)
Comment (dlstadther) [inline luigi/s3.py]: Ah, yes they can

File: luigi/s3.py (hunk 4 above)
Comment (dlstadther) [inline luigi/s3.py]: Oops, oversight...

**Comment by dlstadther** [comment]
@f355 thanks for taking a look! Will fix those two small things

**Comment by jpiper** [comment]
LGTM :+1:

**Comment by dlstadther** [comment]
I will merge this Monday (07/18/2016) morning (US Eastern time) unless I hear otherwise.

**Pull request merged**